addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -1334,77 +1430,90 @@ var SnippetsMenu = Widget.extend({
+                return snippetEditor;
+            }
 
-                // First disable all editors...
-                for (let i = this.snippetEditors.length; i--;) {
-                    const editor = this.snippetEditors[i];
-                    editor.toggleOverlay(false, previewMode);
-                    if (!previewMode && !editorToEnableHierarchy.includes(editor)) {
-                        editor.toggleOptions(false);
-                    }
-                }
-                // ... if no editors are to be enabled, look if any have been
-                // enabled previously by a click
-                if (!editorToEnable) {
-                     editorToEnable = this.snippetEditors.find(editor => editor.isSticky());
-                     previewMode = false;
-                }
-                // ... then enable the right editor
-                if (editorToEnable) {
-                    editorToEnable.toggleOverlay(true, previewMode);
-                    editorToEnable.toggleOptions(true);
+            const snippetEditorHierarchy = [];
+            let currentSnippetEditor = snippetEditor;
+            while (currentSnippetEditor && currentSnippetEditor.$snippetBlock) {
+                snippetEditorHierarchy.push(currentSnippetEditor);
+                currentSnippetEditor = currentSnippetEditor.getParent();
+            }
+
+
+            // First disable all snippet editors...
+            for (const currentSnippetEditor of this.snippetEditors) {
+                currentSnippetEditor.toggleOverlay(false, previewMode);
+                if (!previewMode && !snippetEditorHierarchy.includes(currentSnippetEditor)) {
+                    await currentSnippetEditor.toggleOptions(false);
                 }
+            }
 
-                this._enabledEditorHierarchy = editorToEnableHierarchy;
-                return editorToEnable;
-            });
+            // ... then enable the right snippet editor
+            if (snippetEditor) {
+                snippetEditor.toggleOverlay(true, previewMode);
+                await snippetEditor.toggleOptions(true);
+            }
+
+            this._enabledSnippetEditorsHierarchy = snippetEditorHierarchy;
+            return snippetEditor;
+        });
+    },
+    /**
+     * Disable all snippet editors.
+     *
+     * @private
+     */
+    _enableLastEditor() {
+        this._mutex.exec(() => {
+            // First disable all snippet editors...
+            for (const currentSnippetEditor of this.snippetEditors) {
+                currentSnippetEditor.toggleOverlay(false, false);
+            }
+            // ... then enable the last snippet editor
+            const editorToEnable = this.snippetEditors.find(editor => editor.isSticky());
+            if (editorToEnable) {
+                editorToEnable.toggleOverlay(true, true);
+                editorToEnable.toggleOptions(true);
+            }
```

Comment:
It seems right to me because it's the last instruction in a mutex so it will be awaited when the mutex is. But a second opinion wouldn't hurt. I'll double check with the team on Monday.

Reply:
> it's the last instruction in a mutex so it will be awaited when the mutex is

How would that be possible?

Reply:
There is no await in the code above for that exacte section.
However this is true that `_enableLastEditor` does not return a promise but probably should.

Reply:
What I see is that:
- `_activateSnippet`, in the mutex, awaits for the call to `toggleOptions`
- `_enableLastEditor`, in the mutex, does not awaits for the call to `toggleOptions`... I don't see a reason why it should not

And in the end: why there is not some code sharing between those two functions... but I can't be sure without understanding the purpose of this `_enableLastEditor`

Reply:
I agree with @qsm-odoo's analysis. This is indeed pretty weird.

Reply:
Yup. As mentioned above, there's something fundamentally weird and wrong about `_enableLastEditor` and that should be addressed.